### PR TITLE
[Gen4] Aggregation and Grouping on Operators

### PR DESCRIPTION
**Pull request by harshit-gangal**
## Description

This PR moves the horizon planning (aggregations, group by, order by) to the new operators model.

## Related Issue(s)

- https://github.com/vitessio/vitess/issues/11626

## Checklist

-   [ ] "Backport to:" labels have been added if this change should be back-ported
-   [ ] Tests were added or are not required
-   [ ] Did the new or modified tests pass consistently locally and on the CI
-   [ ] Documentation was added or is not required

### CODE REVIEW

File: go/vt/sqlparser/ast.go [hunk 1]
```diff
@@ -2803,6 +2803,7 @@ type (
 	}
 
 	CountStar struct {
+		inner string // TODO: not sure, but it seems like this is needed. ask @vmg why
```
Comment (harshit-gangal) [inline go/vt/sqlparser/ast.go:2806]: @vmg why is this needed in count(*) function?

File: go/vt/vtgate/planbuilder/operators/aggregation_pushing.go [hunk 2]
```diff
@@ -0,0 +1,315 @@
...
+		Columns:     projections,
+	}
+
+	if !aggregator.Original {
+		// we only keep the root aggregation, if this aggregator was created
+		// by splitting one and pushing under a join, we can get rid of this one
+		return proj, rewrite.NewTree, nil
+	}
+
+	aggregator.Source = proj
```
Comment (harshit-gangal) [inline go/vt/vtgate/planbuilder/operators/aggregation_pushing.go:180]: The reason is not clear here. Adding additional comments will help

File: go/vt/vtgate/executor_select_test.go [hunk 3]
```diff
@@ -3680,7 +3682,7 @@ func TestSelectAggregationNoData(t *testing.T) {
 		{
 			sql:         `select col, count(*) from user group by col limit 2`,
 			sandboxRes:  sqltypes.MakeTestResult(sqltypes.MakeTestFields("col|count(*)", "int64|int64")),
-			expSandboxQ: "select col, count(*), weight_string(col) from `user` group by col, weight_string(col) order by col asc limit :__upper_limit",
+			expSandboxQ: "select col, count(*), weight_string(col) from `user` group by col, weight_string(col) order by col asc",
```
Comment (harshit-gangal) [inline go/vt/vtgate/executor_select_test.go:3685]: I believe this is the right thing to do, it would mean more data but accurate results.

File: go/vt/vtgate/planbuilder/projection.go [hunk 4]
```diff
@@ -42,6 +42,11 @@ var _ logicalPlan = (*projection)(nil)
 
 // WireupGen4 implements the logicalPlan interface
 func (p *projection) WireupGen4(ctx *plancontext.PlanningContext) error {
+	if p.primitive != nil {
+		// already done for this op
+		return p.source.WireupGen4(ctx)
+	}
+
```
Comment (harshit-gangal) [inline go/vt/vtgate/planbuilder/projection.go:49]: not sure why this is done like this

File: go/vt/vtgate/planbuilder/operators/horizon_planning.go [hunk 5]
```diff
@@ -32,25 +36,80 @@ func errHorizonNotPlanned() error {
...
+	defer func() {
+		// If we get the below error.
+		// We will be using old horizon planning.
+		if err == _errHorizonNotPlanned {
+			err = planOffsetsOnJoins(ctx, backup)
+			if err == nil {
+				output = backup
+			}
+		}
+	}()
```
[the remark this answers is not in the document]
Comment (systay) [reply go/vt/vtgate/planbuilder/operators/horizon_planning.go:50]: we need to plan offsets for some stuff even if the rest of horizonPlanning is canceled. this gives us a chance to figure out the offsets of the joincolumns when we had to abort the horizonPlanning process on the operators

File: go/test/endtoend/utils/mysql.go [hunk 6]
```diff
@@ -207,6 +207,10 @@ func compareVitessAndMySQLResults(t *testing.T, query string, vtQr, mysqlQr *sql
 	for _, row := range mysqlQr.Rows {
 		errStr += fmt.Sprintf("%s\n", row)
 	}
+	if vtConn != nil {
+		qr := Exec(t, vtConn, fmt.Sprintf("vexplain plan %s", query))
+		errStr += fmt.Sprintf("query plan: \n%s\n", qr.Rows[0][0].ToString())
+	}
```
Comment (GuptaManan100) [inline go/test/endtoend/utils/mysql.go:213]: Is this for debugging or something you want to actually merge?

File: go/vt/sqlparser/ast.go [hunk 7]
```diff
@@ -2817,6 +2817,7 @@ type (
 	}
 
 	CountStar struct {
+		_ bool // This makes sure that reference equality checks works as expected
```
Comment (GuptaManan100) [inline go/vt/sqlparser/ast.go:2820]: I don't get this 🤷‍♂️

File: go/test/endtoend/utils/mysql.go (hunk 6 above)
Comment (systay) [reply go/test/endtoend/utils/mysql.go:213]: this is a good change that will help. today we only show the plans if the difference is between v3 and gen4, but not when the difference is between mysql and vitess. this change helps us see the plans in the latter case as well

File: go/vt/sqlparser/ast.go (hunk 7 above)
Comment (systay) [reply go/vt/sqlparser/ast.go:2820]: You're correct that this might seem a bit strange at first glance. It's a quirk of Go's handling of empty structs. In Go, two instances of an empty struct are considered identical, which can be problematic when using these as keys in maps. They would be treated as the same key and potentially lead to incorrect map behavior.

Here's a brief example:

```golang
func TestWeirdGo(t *testing.T) {
	type CountStar struct{}

	cs1 := &CountStar{}
	cs2 := &CountStar{}
	require.NotSame(t, cs1, cs2, "what the what!?")
}
```

In the above code, cs1 and cs2, despite being distinct variables, would be treated as the same object.

The solution we employed was to add a dummy field `_ bool` to the otherwise empty struct `CountStar`. This ensures that each instance of `CountStar` is treated as a separate object, even in the context of map keys, such as the semantic state.

**Review by GuptaManan100 — APPROVED** [review]
